drake/systems/plants/RigidBodyManipulator.h

```diff
@@ -97,46 +99,256 @@ class DLLEXPORT_RBM RigidBodyManipulator
   std::string getVelocityName(int velocity_num) const;
   std::string getStateName(int state_num) const;
 
-  template <typename DerivedQ, typename DerivedV>
-  void doKinematics(const MatrixBase<DerivedQ> &q, const MatrixBase<DerivedV> &v, bool compute_gradients = false, bool compute_JdotV = true);
+  template <typename DerivedQ>
+  KinematicsCache<typename DerivedQ::Scalar> doKinematics(const MatrixBase<DerivedQ>& q, int gradient_order) {
+    KinematicsCache<typename DerivedQ::Scalar> ret(bodies, gradient_order);
+    ret.initialize(q);
+    doKinematics(ret);
+    return ret;
+  }
 
-  bool isBodyPartOfRobot(const RigidBody& body, const std::set<int>& robotnum);
+  template <typename DerivedQ, typename DerivedV>
+  KinematicsCache<typename DerivedQ::Scalar> doKinematics(const MatrixBase<DerivedQ>& q, const MatrixBase<DerivedV>& v, int gradient_order = 0, bool compute_JdotV = true) {
+    KinematicsCache<typename DerivedQ::Scalar> ret(bodies, gradient_order);
+    ret.initialize(q, v);
+    doKinematics(ret, compute_JdotV);
+    return ret;
+  }
 
-  double getMass(const std::set<int>& robotnum = RigidBody::defaultRobotNumSet);
+  template <typename Scalar>
+  void doKinematics(KinematicsCache<Scalar>& cache, bool compute_JdotV = false) const {
```

Comment:
do you mean because it’s a class method in the header?  in my mind that is the same as writing the inline keyword.  
> On Sep 4, 2015, at 7:04 AM, Twan Koolen <notifications@github.com> wrote:
> 
> In drake/systems/plants/RigidBodyManipulator.h <https://github.com/RobotLocomotion/drake/pull/1304#discussion_r38741329>:
> 
> >  
> > -  double getMass(const std::set<int>& robotnum = RigidBody::defaultRobotNumSet);
> > +  template <typename Scalar>
> > +  void doKinematics(KinematicsCache<Scalar>& cache, bool compute_JdotV = false) const {
> But then isn't it equally bad form to not add an inline keyword while the compiler does decide to perform inline expansion? And how do I get feedback from the compiler to know if I was right ;-)
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/RobotLocomotion/drake/pull/1304/files#r38741329>.
>